### ISSUE TITLE
Lei Mao's Log Book  Unbiased Estimates in Policy Gradient

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4891]
Lei Mao's Log Book – Unbiased Estimates in Policy Gradient<br>
<br>
https://ift.tt/2Zrzpj6<br>
<br>
Twitter<br>
<br>
<div>
<div> <h3>Introduction</h3> <p>In reinforcement learning policy gradient methods, the goal is usually to maximize $\mathbb{E}[R_t]$, the expected value of return $R$ at time step $t$, where $R_t = \sum_{k=0}^{\infty} \gamma^k r_{t+k}$.</p> <p><br></p> <p>We have parameterized the policy $\pi(a|s; \theta)$ and update the parameter $\theta$ so that $\mathbb{E}[R_t]$ is maximized.</p> <p><br></p> <p>Standard REINFORCE algorithm updates policy in the direction $\nabla_{\theta} R_t \log \pi (a_t|s_t; \theta)$. That is to say, we have a loss function of $ -\log R_t \pi (a_t|s_t; \theta)$ and we are minimizing the loss.</p> <div> <figure><img src="https://leimao.github.io/images/blog/2019-03-15-RL-Policy-Gradient-Unbiase-Estimate/reinforce.png"><figcaption>REINFORCE</figcaption></figure>
</div> <p>But in order to mamixize $\mathbb{E}[R_t]$, why do we have to calculate the the derivatives for $R_t \log \pi (a_t|s_t; \theta)$? It turns out that it is because $\nabla_{\theta} R_t \log \pi (a_t|s_t; \theta)$ is an unbiased estimate of $\nabla_{\theta} \mathbb{E}[R_t]$, i.e., $\mathbb{E}[R_t \nabla_{\theta} \log \pi (a_t|s_t; \theta)] = \nabla_{\theta} \mathbb{E}[R_t]$.</p> <h3>Proof</h3> <p>We use the definition of expected value, because random variable $R_t$ is dependenton $a_t$ and it follows distribution $\pi (a_t|s_t; \theta)$.</p> <p>Then we apply a special case of <a href="https://en.wikipedia.org/wiki/Leibniz_integral_rule">Leibniz Integral Rule</a>, where</p> <p>We then have</p> <p>Because $R_t$ is dependenton $a_t$ but not dependent on $\theta$ (Is it?). We have</p> <p>We apply an identity trick here</p> <p>Because the random variable $R_t \nabla_{\theta} \log \pi (a_t|s_t; \theta)$ also follows distribution $\pi (a_t|s_t; \theta)$, we have</p> <p>Therefore,</p> <p>This concludes the proof.</p> <p><br></p> <h3>Generalization</h3> <p>More generally it is not hard to see that,</p> <h3>Extension</h3> <p>We also use REINFORCE with baseline to reduce tha variance of the estimate of $\nabla_{\theta} \mathbb{E}[R_t]$ while keeping the estimate unbiased.</p> <div> <figure><img src="https://leimao.github.io/images/blog/2019-03-15-RL-Policy-Gradient-Unbiase-Estimate/reinforce.png"><figcaption>REINFORCE with Baseline</figcaption></figure>
</div> <p>In stead of using $\nabla_{\theta} R_t \log \pi (a_t|s_t; \theta)$ as an unbiased estimate of $\nabla_{\theta} \mathbb{E}[R_t]$, we use $\nabla_{\theta} (R_t - v(s_t)) \log \pi (a_t|s_t; \theta)$ as long as $v(s_t)$ does not dependent on $a$.</p> <p><br></p> <p>To see why $\mathbb{E}[(R_t - v(s_t)) \nabla_{\theta} \log \pi (a_t|s_t; \theta)] = \nabla_{\theta} \mathbb{E}[R_t]$, we have</p> <p>Where $\nabla_{\theta} \int v(s_t) \pi (a_t|s_t; \theta) d(a_t) = 0$.</p> <p><br></p> <p>Because $v$ use different parameters other than $\theta$ and it is not dependent on $a$,</p> <h3>Quick Proof to Leibniz Integral Rule</h3> <p>Thanks to the guidance from my friend <a href="https://www.linkedin.com/in/guotu">Guotu Li</a> on this.</p> <p>There is a statement or an assumption we made during the proof, that is $R_t$ is not dependent on $\theta$. It is because in this case $R_t$ is a scala value instead of an output from the policy neural network.</p> <hr>
<footer><div> <img alt="Lei Mao bio photo" src="https://leimao.github.io/images//author_images/Lei-Bio-Medium.jpg"><h3>Lei Mao</h3> <p>Machine Learning, Artificial Intelligence. On the Move.</p> <a href="http://twitter.com/matchaleimao" target="_blank"> Twitter</a> <a href="http://facebook.com/dukeleimao" target="_blank"><i></i> Facebook</a> <a href="http://linkedin.com/in/lei-mao" target="_blank"><i></i> LinkedIn</a> <a href="http://github.com/leimao" target="_blank"><i></i> GitHub</a> <a href="http://scholar.google.com/citations?user=R2VUf7YAAAAJ" target="_blank"><i></i>  G. Scholar</a> <a href="mailto:dukeleimao@gmail.com" target="_blank"><i></i> E-Mail</a> </div> <p><strong>Unbiased Estimates in Policy Gradient</strong> was published on <time datetime="2019-03-15T00:00:00-05:00">March 15, 2019</time> and last modified on <time datetime="2019-03-15">March 15, 2019</time> by <a title="About Lei Mao" href="https://leimao.github.io">Lei Mao</a>.</p> </footer>
</div> </div><br>
<br>
<br>
<br>
<br>
<br>
via Lei Mao's Log Book https://ift.tt/2MtcO1I<br>
<br>
July 31, 2019 at 10:57AM
